src/rules/no-space-in-emphasis.js

```diff
@@ -107,182 +78,97 @@ export default {
-						originalText,
-						checkIndex: endMarker.startIndex - 1,
-						highlightStartIndex: endMarker.startIndex - 2,
-						highlightEndIndex: endMarker.endIndex,
-						removeIndex: nodeStartOffset + endMarker.startIndex - 1,
-						nodeStartLine,
-						nodeStartColumn,
-					});
-				}
-			}
-		}
-
-		/**
-		 * Manager for building a masked character array for a node's direct text content.
-		 * @typedef {{ buffer: string[], startOffset: number }} BufferState
-		 */
-		const bufferManager = {
-			/** @type {BufferState | null} */
-			state: null,
-
-			/**
-			 * Initialize state with a whitespace-masked character buffer for the node.
-			 * @param {Heading|Paragraph|TableCell} node Heading, Paragraph, or TableCell node to enter.
-			 * @returns {void}
-			 */
-			enter(node) {
+		return {
+			"heading, paragraph, tableCell"(
+				/** @type {Heading | Paragraph | TableCell} */ node,
+			) {
 				const [startOffset, endOffset] = sourceCode.getRange(node);
-				this.state = {
-					buffer: new Array(endOffset - startOffset).fill(" "),
-					startOffset,
-				};
-			},
 
-			/**
-			 * Add the content of a Text node into the current buffer at the correct offsets.
-			 * @param {Text} node Text node whose characters will be copied into the buffer.
-			 * @returns {void}
-			 */
-			addText(node) {
-				const start =
-					node.position.start.offset - this.state.startOffset;
-				const text = sourceCode.getText(node);
-				for (let i = 0; i < text.length; i++) {
-					this.state.buffer[start + i] = text[i];
-				}
+				// Initialize `buffer` with a whitespace-masked character array.
+				buffer = new Array(endOffset - startOffset).fill(" ");
 			},
 
-			/**
-			 * Join the character buffer into a masked string, run checks, then clear state.
-			 * @param {Heading|Paragraph|TableCell} node Heading, Paragraph, or TableCell node to exit.
-			 * @returns {void}
-			 */
-			exit(node) {
-				checkEmphasis(node, this.state.buffer.join(""));
-				this.state = null;
-			},
-		};
+			":matches(heading, paragraph, tableCell) > text"(
+				/** @type {Text} */ node,
+			) {
+				const [startOffset, endOffset] = sourceCode.getRange(node);
+				const parentNodeStartOffset = // Parent node can be `Heading`, `Paragraph`, or `TableCell`.
+					sourceCode.getParent(node).position.start.offset;
 
-		return {
-			heading(node) {
-				bufferManager.enter(node);
-			},
-			"heading > text"(node) {
-				bufferManager.addText(node);
-			},
-			"heading:exit"(node) {
-				bufferManager.exit(node);
+				// Add the content of a `Text` node into the current buffer at the correct offsets.
+				for (let i = startOffset; i < endOffset; i++) {
+					buffer[i - parentNodeStartOffset] = sourceCode.text[i];
+				}
 			},
 
-			paragraph(node) {
-				bufferManager.enter(node);
-			},
-			"paragraph > text"(node) {
-				bufferManager.addText(node);
-			},
-			"paragraph:exit"(node) {
-				bufferManager.exit(node);
-			},
+			":matches(heading, paragraph, tableCell):exit"(
```

Comment:
```suggestion
			"heading, paragraph, tableCell:exit"(
```
The `:matches` pseudo-selector is not required.